sys/androidmedia/gstamcvideodec.c

```diff
@@ -1550,101 +1542,102 @@ gst_amc_video_dec_set_format (GstVideoDecoder * decoder,
       }
-      gst_query_unref (query);
     }
-  }
 #else /* Use eglglessink */
-  if (klass->direct_rendering && self->surface == NULL) {
-    self->surface = gst_jni_surface_new (gst_jni_surface_texture_new ());
-    jsurface = self->surface->jobject;
-  }
+    if (klass->direct_rendering && self->surface == NULL) {
+      self->surface = gst_jni_surface_new (gst_jni_surface_texture_new ());
+      jsurface = self->surface->jobject;
+    }
 #endif
 
-  format_string = gst_amc_format_to_string (format);
-  GST_DEBUG_OBJECT (self, "Configuring codec with format: %s surface: %p "
-      "audio session id:%d", format_string, jsurface, self->audio_session_id);
-  g_free (format_string);
-
-  /* We decide that stream is encrypted if we eather received and parsed
-     drm event, eather received crypto ctx from user. It may be not completely correct.
-     Other way - is to base on caps of sinkpad (if they're x-cenc) */
-  if (self->crypto_ctx.mcrypto)
-    self->is_encrypted = TRUE;
+    mime = gst_jni_amc_video_caps_to_mime (state->caps);
+    if (!mime) {
+      GST_ERROR_OBJECT (self, "Failed to convert caps to mime");
+      return FALSE;
+    }
 
-  if (!gst_amc_codec_configure (self->codec, format, jsurface,
-          self->crypto_ctx.mcrypto, 0, self->audio_session_id)) {
-    gst_amc_format_free (format);
-    GST_ERROR_OBJECT (self, "Failed to configure codec");
-    return FALSE;
-  }
+    format = gst_amc_format_new_video (mime, state->info.width,
+        state->info.height);
+    if (!format) {
+      GST_ERROR_OBJECT (self, "Failed to create video format");
+      return FALSE;
+    }
 
-  gst_amc_format_free (format);
+    /* FIXME: This buffer needs to be valid until the codec is stopped again */
+    if (self->codec_data)
+      gst_amc_format_set_buffer (format, "csd-0", self->codec_data);
+
+    /* We decide that stream is encrypted if we eather received and parsed
+       drm event, eather received crypto ctx from user. It may be not completely correct.
+       Other way - is to base on caps of sinkpad (if they're x-cenc) */
+    if (self->crypto_ctx.mcrypto)
+      self->is_encrypted = TRUE;
+
+    format_string = gst_amc_format_to_string (format);
+    GST_DEBUG_OBJECT (self, "Configuring codec with format: %s surface: %p "
+        "audio session id:%d", format_string, jsurface, self->audio_session_id);
+    g_free (format_string);
+
+    if (!gst_amc_codec_configure (self->codec, format, jsurface,
+            self->crypto_ctx.mcrypto, 0, self->audio_session_id)) {
+      GST_ERROR_OBJECT (self, "Failed to configure codec");
+      gst_amc_format_free (format);
+      return FALSE;
+    }
 
-  if (!gst_amc_codec_start (self->codec)) {
-    GST_ERROR_OBJECT (self, "Failed to start codec");
-    return FALSE;
-  }
+    if (!gst_amc_codec_start (self->codec)) {
+      GST_ERROR_OBJECT (self, "Failed to start codec");
+      gst_amc_format_free (format);
+      return FALSE;
+    }
+    gst_amc_format_free (format);
 
```

Comment:
I don't see why not freeing format before calling  gst_amc_codec_start() as previously. Did you find it is still required?

Reply:
Done.